CommandLine/emake/Server.cpp

```diff
@@ -15,23 +15,51 @@
 #include <boost/filesystem.hpp>
 
 #include <memory>
+#include <future>
+#include <sstream>
 
 using namespace grpc;
 using namespace buffers;
 
 class CompilerServiceImpl final : public Compiler::Service {
   public:
-  explicit CompilerServiceImpl(EnigmaPlugin& plugin, OptionsParser& options): plugin(plugin), options(options) {}
+  explicit CompilerServiceImpl(EnigmaPlugin& plugin, OptionsParser& options):
+    plugin(plugin), options(options) {
+      // create output streams and redirect standard output to them
+      // this way we can read from the stream and forward to the client
+      std::cout.rdbuf(outstm.rdbuf());
+      std::cerr.rdbuf(errstm.rdbuf());
+    }
 
   Status CompileBuffer(ServerContext* /*context*/, const CompileRequest* request, ServerWriter<CompileReply>* writer) override {
-    plugin.BuildGame(const_cast<buffers::Game*>(&request->game()), emode_run, request->name().c_str());
+    // block until the last CompileBuffer rpc finishes
+    if (CompileBufferFuture.valid()) CompileBufferFuture.get();
+    // use lambda capture to contain compile logic
+    auto fnc = [=] {
+      plugin.BuildGame(const_cast<buffers::Game*>(&request->game()), emode_run, request->name().c_str());
+    };
+    // asynchronously launch the compile request
+    CompileBufferFuture = std::async(fnc);

```

Comment:
AFAICT, all you're doing here is threading the compiler. This will not rate limit it unless you have a single-worker pool somewhere to handle these requests. As discussed over Discord, I think it would be wise to enable multiple compilers to run, in which case, use a worker pool of size three (or something reasonable) and handle the queue piling up.

Generally, if the queue piles up, the proper strategy is to reject the oldest request first (the next person in line). The reason for that is, it's better to have a user hit compile again (and possibly be load balanced to a less busy instance) than to have users in that queue keep waiting on a process that will never finish or that will finish eventually but is no longer even wanted.

Reply:
Right, but technically it was rate limited to one job, because I was calling `get()` on the previous future first which is blocking. With 36dac03f55f24fc0d46082042cfd1de594e6d234 I am also still blocking until the future completes during which I loop and send back log messages.

Please answer my question above about whether I should have the log method be a callback which the plugin implements and then have the logging queue stored in the plugin instance.